File: src/main/java/leet_code/GreatestCommonDivisorStrings.java
package leet_code;

public class GreatestCommonDivisorStrings {

//    For two strings s and t, we say "t divides s" if and only if s = t + ... + t (i.e., t is concatenated with itself one or more times).
//
//    Given two strings str1 and str2, return the largest string x such that x divides both str1 and str2.
//
//
//
//    Example 1:
//
//    Input: str1 = "ABCABC", str2 = "ABC"
//    Output: "ABC"
//    Example 2:
//
//    Input: str1 = "ABABAB", str2 = "ABAB"
//    Output: "AB"
//    Example 3:
//
//    Input: str1 = "LEET", str2 = "CODE"
//    Output: ""

    public static void main(String[] args) {


    }


    public static String gcdOfStrings(String str1, String str2) {
        if (!(str1 + str2).equals(str2 + str1)) return "";

        int len1 = str1.length();
        int len2 = str2.length();

        int gcd = 1;

        int min = Math.min(len1, len2);

        for (int i = 2; i <= min; i++) {
            if (len1 % i == 0 && len2 % i == 0) {
                gcd = i;
            }
        }
        return str2.substring(0, gcd);
    }
}

File: src/main/java/leet_code/till_april/GreatestCommonDivisorStrings.go
package main

import "fmt"

func main() {
	fmt.Println(gcdOfStrings("ABCABC", "ABC"))
	fmt.Println(gcdOfStrings("ABABAB", "ABAB"))
}

func gcdOfStrings(str1 string, str2 string) string {
	if (str1 + str2) != str2+str1 {
		fmt.Println("worked i")
		return ""
	}
	minL := minGcd(len(str1), len(str2))
	maxL := maxGcd(len(str1), len(str2))
	for minL > 0 {
		maxL, minL = minL, maxL%minL
	}
	return str1[0:maxL]

}

func minGcd(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func maxGcd(a, b int) int {
	if a > b {
		return a
	}
	return b
}
